File: LeetCode/Java/src/solutions/p0109_ConvertSortedListtoBinarySearchTree/P0109.java
package solutions.p0109_ConvertSortedListtoBinarySearchTree;

import auxiliary.ListNode;
import auxiliary.TreeNode;

public class P0109 {
    public TreeNode sortedListToBST(ListNode head) {
        /*
         * Solution v1: Two poniters to find middle node
         *
         * Runtime: 0 ms @ (beats) 100%
         * Memory Usage: 43.7 MB @ (beats) 44.80%
         */

        if (head == null) {
            return null;
        } else if (head.next == null) {
            return new TreeNode(head.val);
        }

        ListNode slow = head;
        ListNode fast = head.next.next;

        while (fast != null && fast.next != null) {
            slow = slow.next;
            fast = fast.next.next;
        }

        TreeNode tree = new TreeNode(slow.next.val);

        tree.right = this.sortedListToBST(slow.next.next);

        slow.next = null;

        tree.left = this.sortedListToBST(head);

        return tree;
    }
}

File: LeetCode/Python/solutions/p0109.py
from typing import Optional

from Python.auxiliary.ListNode import ListNode
from Python.auxiliary.TreeNode import TreeNode


class Solution:
    def sortedListToBST(self, head: Optional[ListNode]) -> Optional[TreeNode]:
        #
        # Solution v1: Recursive
        #
        # Runtime: 141 ms @ (beats) 15.86%
        # Memory Usage: 20.3 MB @ (beats) 26.41%
        #
        # def helper(vals: list) -> Optional[TreeNode]:
        #     if not vals:
        #         return None
        #
        #     tree = TreeNode()
        #     mid = len(vals) // 2
        #     tree.val = vals[mid]
        #     tree.left = helper(vals[:mid])
        #     tree.right = helper(vals[mid + 1:])
        #     return tree
        #
        # nodes = []
        # while head:
        #     nodes.append(head.val)
        #     head = head.next
        #
        # return helper(nodes)

        #
        # Solution v2: Two pointer to find middle node
        #
        # Runtime: 121 ms @ (beats) 82.47%
        # Memory Usage: 17.7 MB @ (beats) 76.91%
        #

        if not head:
            return None
        elif not head.next:
            return TreeNode(head.val)

        slow = head
        fast = head.next.next

        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next

        tree = TreeNode(slow.next.val)

        tree.right = self.sortedListToBST(slow.next.next)

        # cut the ListNode
        slow.next = None

        tree.left = self.sortedListToBST(head)

        return tree

